pkg/ingester/wal.go

```diff
@@ -384,35 +383,89 @@ func recoverFromWAL(ingester *Ingester) (err error) {
 		seriesCache[i] = make(map[string]map[uint64]*memorySeries)
 	}
 
-	if idx >= 0 {
-		// Checkpoint exists.
-		level.Info(util.Logger).Log("msg", "recovering from checkpoint", "checkpoint", lastCheckpointDir)
-		start := time.Now()
-		if err := processCheckpoint(lastCheckpointDir, userStates, nWorkers, stateCache, seriesCache); err != nil {
-			return err
-		}
-		elapsed := time.Since(start)
-		level.Info(util.Logger).Log("msg", "recovered from checkpoint", "time", elapsed.String())
-	} else {
-		level.Info(util.Logger).Log("msg", "no checkpoint found")
+	level.Info(util.Logger).Log("msg", "recovering from checkpoint")
+	start := time.Now()
+	userStates, idx, err := processCheckpointWithRepair(walDir, ingester, nWorkers, stateCache, seriesCache)
+	if err != nil {
+		return err
 	}
+	elapsed := time.Since(start)
+	level.Info(util.Logger).Log("msg", "recovered from checkpoint", "time", elapsed.String())
+	defer func() {
+		if returnErr == nil {
```

Comment:
`returnErr` is never valued

Reply:
Not sure I understand. As I have `returnErr` as a return variable, any error that is returned later will be bound to it.

Reply:
Wow, you're right: https://play.golang.org/p/zeWCaztXUGh

But again it's non-obvious to me and Marco and I assume it'll be the same for others. Can we somehow refactor to use `returnErr` everywhere? 

Or not put this in a defer? I think it'll be clearer and more obvious if we manually set it and not put it in a defer.

Reply:
Okay